src/Host/Client/Impl/Session/RSession.cs

```diff
@@ -262,34 +266,31 @@ internal sealed class RSession : IRSession, IRCallbacks {
 
             await _initializationTcs.Task;
             initializationCts.Dispose();
+            _stopHostLock.EnqueueReset();
             lockToken.Set();
         }
 
-        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionSwitchBrokerTransaction StartSwitchingBroker() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
-        public IRSessionReconnectTransaction StartReconnecting() {
-            _disposeToken.ThrowIfDisposed();
-            return new BrokerTransaction(this);
-        }
+        public IRSessionReconnectTransaction StartReconnecting() => !_disposeToken.IsDisposed ? new BrokerTransaction(this) : null;
 
         public async Task StopHostAsync() {
             _disposeToken.ThrowIfDisposed();
```

Comment:
I think that `Dispose` as cancellation for any in-flight async operations makes sense. Making it wait on in-flight operations would be very surprising, since that's not normally how `Dispose` behaves; and some of those (e.g. evals) can be pretty long running, too, introducing undesired blocking.